### PR TITLE
Rewrite bucket-generating logic

### LINKED ISSUE #397: The arrows on the bucket bar double count close together annotations
### Steps to reproduce

1. Create an annotation
2. Create a second annotation close to the first. The highlighted texts of the two annotations don't have to be overlapping, but they need to be very close together

### Expected behaviour

There are two annotations, so I'd expect the sum of the numbers shown in the arrows in the bucket bar to be 2.

### Actual behaviour

The numbers sum to more than 2 because some of the annotations are double-counted. With 2 annotations it can be 3 or 4.

peek 2017-05-23 11-04

### PR DESCRIPTION
**Pull request by lyzadanger**
This ~~draft PR demonstrates an approach to refactoring~~ reimplements the logic used to compute and position anchor buckets in the bucket bar. ~~It does not yet have tests.~~

It is my hope that this re-implementation will be more comprehensible to future devs, provide a clearer separation of concerns and more clearly express the desired outcome of the bucketing logic.

It:

* Removes the weird empty extra buckets
* Redefines types in `buckets` util in a way that feels more intuitive
* Moves some positioning "logic" into CSS
* Removes the need for `BucketBar` to maintain (and export) constants
* Fixes [this bug](https://github.com/hypothesis/client/issues/397)

The resulting buckets are very close to what the previous algorithm created, minus the knack for sometimes putting anchors into multiple buckets (bug #397). 

There's probably one more follow-up pass to rectify a few lingering...interesting features of the code...but this is feeling better...

Fixes https://github.com/hypothesis/client/issues/2618
Fixes #397

### CODE REVIEW

File: src/annotator/plugin/bucket-bar.js
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:111]: Ahhhhh this feels better.

File: src/annotator/plugin/bucket-bar.js
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:210]: Ahhh, this also feels better.

File: src/annotator/plugin/bucket-bar.js
Comment (lyzadanger) [inline src/annotator/plugin/bucket-bar.js:203]: TODO: Make these methods unnecessary

File: src/annotator/util/buckets.js
Comment (lyzadanger) [inline src/annotator/util/buckets.js:41]: Why YES, I'm serious, this sub-pixel value makes a difference in label alignment within the upper-navigation bucket. I need to get to the bottom of it.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [inline src/annotator/util/buckets.js:207]: So, basically, this seems to be the crux of the entire algorithm as I understand it. It seems quite simple, but seems to do the proverbial trick so far as I can ascertain.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [inline src/annotator/util/buckets.js:214]: I'm hoping that typedef documentation is clear enough to not warrant line-by-line commenting here, but I can add it!

File: src/styles/annotator/bucket-bar.scss
Comment (lyzadanger) [inline src/styles/annotator/bucket-bar.scss:38]: This entire SASS module needs a little TLC, but for now: there's no need to make `BucketBar` handle top margins.

File: src/annotator/util/buckets.js
Comment (robertknight) [inline src/annotator/util/buckets.js:127]: You could write this function as one filter/map/sort chain. I'd be fine with it either way.

File: src/styles/annotator/bucket-bar.scss
Comment (robertknight) [reply src/styles/annotator/bucket-bar.scss:38]: Can you expand on this?

File: src/annotator/plugin/bucket-bar.js
Comment (robertknight) [inline src/annotator/plugin/bucket-bar.js:174]: `_buildTabs` is a misleading name IMO as this function merely updates the properties of existing tab elements, which must have already been constructed by `_update`. Since `_buildTab` is only called in one place at the end of `_update`, I think the code would be just as readable if the body of this method were moved into the `_update` function above - and it would likely make other changes easier.

**Review by robertknight — COMMENTED** [review]
This is a great improvement! I left some minor feedback on the details.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [reply src/annotator/util/buckets.js:156]: Yeah, this current design has a bit o' a wee smell. I'll refactor.

File: src/annotator/plugin/bucket-bar.js
Comment (lyzadanger) [reply src/annotator/plugin/bucket-bar.js:174]: This one is unchanged from the legacy implementation of this module. I'll catch it on my next pass which will update DOM details (I want to attack the a11y issues here).

File: src/annotator/util/buckets.js
Comment (lyzadanger) [reply src/annotator/util/buckets.js:17]: Another option is to leave the `top` and `bottom` properties on `Bucket` objects such that `WorkingBucket` isn't a thing...i.e. we just work with `Bucket`s. Those two properties aren't used elsewhere, though, obviously.

**Comment by lyzadanger** [comment]
@robertknight I've pushed a commit that integrates your initial feedback, most materially fixing the ickiness of the internal bucket-building function formerly known as `_nextBucket`

File: src/styles/annotator/bucket-bar.scss
Comment (lyzadanger) [reply src/styles/annotator/bucket-bar.scss:39]: See updated comments. The whole approach to styling needs an overhaul; it's in my TODO list.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [reply src/annotator/util/buckets.js:17]: I had sorta forgotten that typechecking doesn't mind _extra_ properties, so I'm more flexibly using `WorkingBuckets` where `Buckets` are expected in some convenient cases. See updated code.

File: src/annotator/util/buckets.js
Comment (robertknight) [inline src/annotator/util/buckets.js:152]: Can you remove the underscore prefix here. Perhaps it was needed in an earlier iteration of the code but it isn't needed any longer

This helper could be moved to the top level of the module as it doesn't capture any variables - I leave that up to you.

File: src/annotator/util/buckets.js
Comment (robertknight) [inline src/annotator/util/buckets.js:198]: Since the `anchorPositions` collection is sorted by the `top` property, this condition will always be false. The anchor currently being visited will never have a lower `top` property than the earlier anchor which set `currentBucket.top`. Hence you can eliminate the `updatedTop` variable.

File: src/annotator/util/test/buckets-test.js
Comment (robertknight) [inline src/annotator/util/test/buckets-test.js:166]:
```suggestion
    it('sorts anchors by top position', () => {
```

Just a suggestion. I found the previous text slightly awkward.

File: src/styles/annotator/bucket-bar.scss
Comment (robertknight) [inline src/styles/annotator/bucket-bar.scss:39]: This property duplicates the `margin-top` rule a few lines above.

File: src/styles/annotator/bucket-bar.scss
Comment (robertknight) [inline src/styles/annotator/bucket-bar.scss:100]: "intended position" means "the dynamically set `top` style property"?

File: src/styles/annotator/bucket-bar.scss
Comment (robertknight) [inline src/styles/annotator/bucket-bar.scss:99]: The "22px" value comes from the height of the `.upper` element plus the triangle created by the `::before` and `::after` pseudo-selectors?

**Review by robertknight — APPROVED** [review]
LGTM. This is a huge improvement over the version on `master`. I requested some minor changes, after which this is good to merge. Give me a shout if you have a query about any of those.

The existing CSS structure is fiddly to understand due to use of `::before` and `::after` to create the arrows and the amount of overriding of properties that happens to style the `.upper` and `.lower` indicators. At least it is self-contained though and we can revisit that later.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [reply src/annotator/util/buckets.js:152]: I'll kill the underscore, no prob. I'd like to keep this function here as it's only relevant to the iteration beneath it and I don't want a future dev to have to do too much searching through the module to find it.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [reply src/annotator/util/buckets.js:198]: Ah, yes, as background, this logic was written before the API surface between `BucketBar` and `buckets` was rewritten to be a single function; because this function took the "points" provided by `BucketBar`, it couldn't entirely rely on them to be sorted...

File: src/styles/annotator/bucket-bar.scss
Comment (lyzadanger) [reply src/styles/annotator/bucket-bar.scss:100]: Yes, I'll clarify.

File: src/styles/annotator/bucket-bar.scss
Comment (lyzadanger) [reply src/styles/annotator/bucket-bar.scss:39]: Ah, thanks for spotting. This SASS module is...chattery :)

File: src/styles/annotator/bucket-bar.scss
Comment (lyzadanger) [reply src/styles/annotator/bucket-bar.scss:99]: Correct. The entire element takes up 22px of space, vertically.

File: src/annotator/util/buckets.js
Comment (lyzadanger) [reply src/annotator/util/buckets.js:198]: Fascinatingly, this change breaks tests. I'll have to investigate.

Comment (lyzadanger) [reply src/annotator/util/buckets.js:198]: Pilot error, ignore.

**Pull request merged**
